federated_learning/run_toxicity.py

```diff
@@ -0,0 +1,223 @@
+        model_fn,
+        client_optimizer_fn=lambda: tf.keras.optimizers.Adam(clientlr),
+        server_optimizer_fn=lambda: tf.keras.optimizers.Adam(learning_rate=lr))
+
+    evaluation = tff.learning.build_federated_evaluation(model_fn)
+
+    all_test_acc = []
+    all_test_loss = []
+    all_test_auc = []
+    current_time = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
+    logdir = os.path.join('logs', 'toxicity', current_time)
+    writer = tf.summary.create_file_writer(logdir)
+
+    for k in range(subsets):
+        state = trainer.initialize()
+        test_data_idx = k
+        val_data_idx = k - 1 if k == (subsets - 1) else k + 1
+        test_data = [all_data[test_data_idx], ]
+        val_data = [all_data[val_data_idx], ]
+        train_data = [d for idx, d in enumerate(all_data) if not idx in [
+            test_data_idx, val_data_idx]]
+        logger.debug(f'{k} round ->')
+
+        for round_num in range(rounds):
+            state, metrics = trainer.next(state, train_data)
+            train_acc = metrics['train']["binary_accuracy"]
+            train_loss = metrics['train']["loss"]
+            train_auc = metrics['train']["auc"]
+            logger.debug(f'train, round, loss, acc, auc  ===> {round_num:03d}, {train_loss:7.5f}, '
+                         f'{train_acc:7.5f}, {train_auc:7.5f},')
+            val_metrics = evaluation(state.model, val_data)
+            val_acc = val_metrics["binary_accuracy"]
+            val_loss = val_metrics["loss"]
+            val_auc = val_metrics["auc"]
+            logger.debug(f'val, round, loss, acc, auc ===> {round_num:03d}, {val_loss:7.5f}, '
+                         f'{val_acc:7.5f}, {val_auc:7.5f},')
+            with writer.as_default():
+                tf.summary.scalar(f'train_loss{k}', train_loss, step=round_num)
+                tf.summary.scalar(f'train_auc{k}', train_auc, step=round_num)
+                tf.summary.scalar(f'val_loss{k}', val_loss, step=round_num)
+                tf.summary.scalar(f'val_auc{k}', val_auc, step=round_num)
+
+        test_metrics = evaluation(state.model, test_data)
+        test_acc = test_metrics["binary_accuracy"]
+        test_loss = test_metrics["loss"]
+        test_auc = test_metrics["auc"]
+        with writer.as_default():
+            tf.summary.scalar(f'test_loss{k}', test_loss, step=round_num)
+            tf.summary.scalar(f'test_auc{k}', test_auc, step=round_num)
+        all_test_acc.append(test_acc)
+        all_test_loss.append(test_loss)
+        all_test_auc.append(test_auc)
+        logger.debug(f'test, round, loss, acc, auc ===> {k}, {round_num:03d}, {test_loss:7.5f}, '
+                     f' {test_acc:7.5f}, {test_auc:7.5f},')
+    logger.debug(f'{clients} >>>> all_test_acc = {all_test_acc}')
+    logger.debug(f'{clients} >>>> all_test_loss = {all_test_loss}')
+    logger.debug(f'{clients} >>>> all_test_auc = {all_test_auc}')
+
+
+def normal_learning(rounds, epochs, batchsize, lr, model, dataset_name):
```

Comment:
For training without federated learning. This code ran with the following log. The loss and accuracy is not good at all, but I could not judge this is a problem of dataset, or could be improved by some tuning.
```
Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
features (InputLayer)           [(None, 250, 100)]   0                                            
__________________________________________________________________________________________________
adjs (InputLayer)               [(None, 1, 250, 250) 0                                            
__________________________________________________________________________________________________
graph_conv_fl (GraphConvFL)     (None, 250, 64)      6464        features[0][0]                   
__________________________________________________________________________________________________
re_lu (ReLU)                    (None, 250, 64)      0           graph_conv_fl[0][0]              
__________________________________________________________________________________________________
graph_conv_fl_1 (GraphConvFL)   (None, 250, 64)      4160        re_lu[0][0]                      
__________________________________________________________________________________________________
re_lu_1 (ReLU)                  (None, 250, 64)      0           graph_conv_fl_1[0][0]            
__________________________________________________________________________________________________
graph_gather (GraphGather)      (None, 64)           0           re_lu_1[0][0]                    
__________________________________________________________________________________________________
dense (Dense)                   (None, 1)            65          graph_gather[0][0]               
==================================================================================================
Total params: 10,689
Trainable params: 10,689
Non-trainable params: 0
__________________________________________________________________________________________________
Epoch 1/10
116/116 [==============================] - 4s 35ms/step - loss: 977.4376 - binary_accuracy: 0.9587 - auc: 0.9598
Epoch 2/10
116/116 [==============================] - 4s 37ms/step - loss: 0.8968 - binary_accuracy: 0.7163 - auc: 0.7719   
Epoch 3/10
116/116 [==============================] - 4s 35ms/step - loss: 0.9682 - binary_accuracy: 0.6904 - auc: 0.7160   
Epoch 4/10
116/116 [==============================] - 4s 35ms/step - loss: 0.9299 - binary_accuracy: 0.6731 - auc: 0.7187   
Epoch 5/10
116/116 [==============================] - 4s 35ms/step - loss: 0.9177 - binary_accuracy: 0.6731 - auc: 0.7187   
Epoch 6/10
116/116 [==============================] - 4s 34ms/step - loss: 0.9128 - binary_accuracy: 0.6731 - auc: 0.7140   
Epoch 7/10
116/116 [==============================] - 4s 35ms/step - loss: 0.9084 - binary_accuracy: 0.6731 - auc: 0.7139   
Epoch 8/10
116/116 [==============================] - 4s 35ms/step - loss: 0.9071 - binary_accuracy: 0.6725 - auc: 0.7112   
Epoch 9/10
116/116 [==============================] - 4s 35ms/step - loss: 0.9036 - binary_accuracy: 0.6731 - auc: 0.7133   
Epoch 10/10
116/116 [==============================] - 4s 35ms/step - loss: 0.9042 - binary_accuracy: 0.6725 - auc: 0.7113   
1851/1851 [==============================] - 2s 1ms/step - loss: 2.0652 - binary_accuracy: 0.4706 - auc: 0.5000
```

Reply:
What learning rate did you use in your simulation? Default parameter is 0.2. It seems that the learning rate is too high for Adam optimizer. If so, could you try to use 0.001 as a learning rate?

Reply:
Yes, I ran with the default parameters. I tried with smaller lr, and obtained the following log. 
The accuracy and loss seem to be decreasing simultaneously. Does this mean setting around them is someting wrong ?
- lr = 0.001
```
Epoch 1/20
116/116 [==============================] - 4s 35ms/step - loss: 0.5275 - binary_accuracy: 0.9011 - auc: 0.9499
Epoch 2/20
116/116 [==============================] - 4s 35ms/step - loss: 1.4807 - binary_accuracy: 0.6774 - auc: 0.7246    
Epoch 3/20
116/116 [==============================] - 4s 35ms/step - loss: 0.9248 - binary_accuracy: 0.5834 - auc: 0.6476   
Epoch 4/20
116/116 [==============================] - 4s 37ms/step - loss: 0.8844 - binary_accuracy: 0.3540 - auc: 0.3127   
Epoch 5/20
116/116 [==============================] - 4s 38ms/step - loss: 0.6760 - binary_accuracy: 0.4988 - auc: 0.4538
Epoch 6/20
116/116 [==============================] - 5s 46ms/step - loss: 0.7455 - binary_accuracy: 0.4431 - auc: 0.3807   
Epoch 7/20
116/116 [==============================] - 5s 39ms/step - loss: 0.6902 - binary_accuracy: 0.5504 - auc: 0.4056
Epoch 8/20
116/116 [==============================] - 5s 40ms/step - loss: 0.6919 - binary_accuracy: 0.5293 - auc: 0.4059
Epoch 9/20
116/116 [==============================] - 4s 39ms/step - loss: 0.6885 - binary_accuracy: 0.5550 - auc: 0.4292
Epoch 10/20
116/116 [==============================] - 5s 39ms/step - loss: 0.6885 - binary_accuracy: 0.5531 - auc: 0.4354
Epoch 11/20
116/116 [==============================] - 4s 38ms/step - loss: 0.6876 - binary_accuracy: 0.5542 - auc: 0.4411
Epoch 12/20
116/116 [==============================] - 4s 38ms/step - loss: 0.6870 - binary_accuracy: 0.5542 - auc: 0.4498
Epoch 13/20
116/116 [==============================] - 4s 38ms/step - loss: 0.6860 - binary_accuracy: 0.5555 - auc: 0.4447
Epoch 14/20
116/116 [==============================] - 4s 38ms/step - loss: 0.6904 - binary_accuracy: 0.5534 - auc: 0.4495
Epoch 15/20
116/116 [==============================] - 5s 39ms/step - loss: 0.6873 - binary_accuracy: 0.5534 - auc: 0.4362
Epoch 16/20
116/116 [==============================] - 4s 37ms/step - loss: 0.6852 - binary_accuracy: 0.5577 - auc: 0.4681
Epoch 17/20
116/116 [==============================] - 4s 36ms/step - loss: 0.6843 - binary_accuracy: 0.5615 - auc: 0.4682
Epoch 18/20
116/116 [==============================] - 4s 36ms/step - loss: 0.5156 - binary_accuracy: 0.8065 - auc: 0.7493
Epoch 19/20
116/116 [==============================] - 5s 40ms/step - loss: 1.0800 - binary_accuracy: 0.7630 - auc: 0.8142
Epoch 20/20
116/116 [==============================] - 5s 40ms/step - loss: 1.0865 - binary_accuracy: 0.3964 - auc: 0.5747
1851/1851 [==============================] - 2s 1ms/step - loss: 0.6874 - binary_accuracy: 0.5505 - auc: 0.5287
```
- lr = 0.0001
```
Epoch 1/20
116/116 [==============================] - 4s 35ms/step - loss: 0.6281 - binary_accuracy: 0.8030 - auc: 0.8933
Epoch 2/20
116/116 [==============================] - 4s 36ms/step - loss: 0.9751 - binary_accuracy: 0.6115 - auc: 0.6697   
Epoch 3/20
116/116 [==============================] - 4s 36ms/step - loss: 0.8886 - binary_accuracy: 0.6058 - auc: 0.6732   
Epoch 4/20
116/116 [==============================] - 4s 35ms/step - loss: 0.8705 - binary_accuracy: 0.6052 - auc: 0.6696   
Epoch 5/20
116/116 [==============================] - 4s 36ms/step - loss: 0.8599 - binary_accuracy: 0.6039 - auc: 0.6639   
Epoch 6/20
116/116 [==============================] - 5s 41ms/step - loss: 0.8553 - binary_accuracy: 0.5936 - auc: 0.6587   
Epoch 7/20
116/116 [==============================] - 4s 35ms/step - loss: 0.8441 - binary_accuracy: 0.5925 - auc: 0.6566   
Epoch 8/20
116/116 [==============================] - 4s 37ms/step - loss: 0.8432 - binary_accuracy: 0.5934 - auc: 0.6496   
Epoch 9/20
116/116 [==============================] - 4s 36ms/step - loss: 0.8351 - binary_accuracy: 0.5877 - auc: 0.6465   
Epoch 10/20
116/116 [==============================] - 4s 36ms/step - loss: 0.8293 - binary_accuracy: 0.5855 - auc: 0.6421   
Epoch 11/20
116/116 [==============================] - 4s 37ms/step - loss: 0.8214 - binary_accuracy: 0.5842 - auc: 0.6382   
Epoch 12/20
116/116 [==============================] - 4s 37ms/step - loss: 0.8180 - binary_accuracy: 0.5769 - auc: 0.6318   
Epoch 13/20
116/116 [==============================] - 4s 37ms/step - loss: 0.8153 - binary_accuracy: 0.5680 - auc: 0.6269   
Epoch 14/20
116/116 [==============================] - 5s 42ms/step - loss: 0.8082 - binary_accuracy: 0.5707 - auc: 0.6264   
Epoch 15/20
116/116 [==============================] - 4s 38ms/step - loss: 0.8139 - binary_accuracy: 0.5712 - auc: 0.6192   
Epoch 16/20
116/116 [==============================] - 4s 37ms/step - loss: 0.8101 - binary_accuracy: 0.5601 - auc: 0.6098   
Epoch 17/20
116/116 [==============================] - 4s 38ms/step - loss: 0.8001 - binary_accuracy: 0.5569 - auc: 0.6091   
Epoch 18/20
116/116 [==============================] - 5s 39ms/step - loss: 0.7980 - binary_accuracy: 0.5561 - auc: 0.6051   
Epoch 19/20
116/116 [==============================] - 5s 40ms/step - loss: 0.7970 - binary_accuracy: 0.5520 - auc: 0.6000   
Epoch 20/20
116/116 [==============================] - 4s 37ms/step - loss: 0.7994 - binary_accuracy: 0.5485 - auc: 0.5928   
1851/1851 [==============================] - 2s 1ms/step - loss: 1.0443 - binary_accuracy: 0.4711 - auc: 0.5141
```

Reply:
Sorry, I may have figured out the problem. 
`buffer_size` for `tf.Dataset.shuffle` was so small that dataset is splitted into train, test, and val in a biased way.
By setting `buffer_size` equals to dataset length, I got the following. 
```
Epoch 1/20
116/116 [==============================] - 4s 36ms/step - loss: 0.7816 - binary_accuracy: 0.5512 - auc: 0.5683
Epoch 2/20
116/116 [==============================] - 4s 35ms/step - loss: 0.6796 - binary_accuracy: 0.5766 - auc: 0.6119
Epoch 3/20
116/116 [==============================] - 4s 35ms/step - loss: 0.6718 - binary_accuracy: 0.5863 - auc: 0.6304
Epoch 4/20
116/116 [==============================] - 4s 35ms/step - loss: 0.6586 - binary_accuracy: 0.6050 - auc: 0.6504
Epoch 5/20
116/116 [==============================] - 4s 35ms/step - loss: 0.6562 - binary_accuracy: 0.6131 - auc: 0.6533
Epoch 6/20
116/116 [==============================] - 4s 35ms/step - loss: 0.6610 - binary_accuracy: 0.6177 - auc: 0.6543
Epoch 7/20
116/116 [==============================] - 4s 35ms/step - loss: 0.6474 - binary_accuracy: 0.6266 - auc: 0.6699
Epoch 8/20
116/116 [==============================] - 4s 34ms/step - loss: 0.6447 - binary_accuracy: 0.6350 - auc: 0.6800
Epoch 9/20
116/116 [==============================] - 4s 35ms/step - loss: 0.6371 - binary_accuracy: 0.6406 - auc: 0.6892
Epoch 10/20
116/116 [==============================] - 4s 37ms/step - loss: 0.6385 - binary_accuracy: 0.6463 - auc: 0.6864
Epoch 11/20
116/116 [==============================] - 4s 36ms/step - loss: 0.6360 - binary_accuracy: 0.6493 - auc: 0.6943
Epoch 12/20
116/116 [==============================] - 4s 36ms/step - loss: 0.6350 - binary_accuracy: 0.6485 - auc: 0.6931
Epoch 13/20
116/116 [==============================] - 4s 35ms/step - loss: 0.6359 - binary_accuracy: 0.6390 - auc: 0.6933
Epoch 14/20
116/116 [==============================] - 4s 36ms/step - loss: 0.6276 - binary_accuracy: 0.6566 - auc: 0.7094
Epoch 15/20
116/116 [==============================] - 4s 35ms/step - loss: 0.6444 - binary_accuracy: 0.6447 - auc: 0.6883
Epoch 16/20
116/116 [==============================] - 4s 37ms/step - loss: 0.6315 - binary_accuracy: 0.6550 - auc: 0.7031
Epoch 17/20
116/116 [==============================] - 4s 36ms/step - loss: 0.6295 - binary_accuracy: 0.6539 - auc: 0.7039
Epoch 18/20
116/116 [==============================] - 4s 36ms/step - loss: 0.6214 - binary_accuracy: 0.6644 - auc: 0.7138
Epoch 19/20
116/116 [==============================] - 4s 36ms/step - loss: 0.6194 - binary_accuracy: 0.6695 - auc: 0.7197
Epoch 20/20
116/116 [==============================] - 4s 36ms/step - loss: 0.6229 - binary_accuracy: 0.6647 - auc: 0.7106
1851/1851 [==============================] - 2s 1ms/step - loss: 0.6236 - binary_accuracy: 0.6667 - auc: 0.7216
```

Reply:
'NTP_PubChem_Bench.20201106" is used(this is set by default)

Reply:
Thanks.